diffkemp/simpll/PatternSet.h

```diff
@@ -0,0 +1,227 @@
+    }
+};
+} // namespace std
+
+// Define a hash function for instruction difference patterns.
+namespace std {
+template <> struct hash<InstPattern> {
+    std::size_t operator()(const InstPattern &Pat) const noexcept {
+        return std::hash<Pattern>()(Pat);
+    }
+};
+} // namespace std
+
+// Define a hash function for value difference patterns.
+namespace std {
+template <> struct hash<ValuePattern> {
+    std::size_t operator()(const ValuePattern &Pat) const noexcept {
+        return std::hash<Pattern>()(Pat);
+    }
+};
+} // namespace std
+
+/// Compares difference patterns against functions, possibly eliminating reports
+/// of prior semantic differences.
+class PatternSet {
+  public:
+    /// Default DiffKemp prefix for all pattern information.
+    static const std::string DefaultPrefix;
+    /// Prefix for the left (old) side of difference patterns.
+    static const std::string PrefixL;
+    /// Prefix for the right (new) side of difference patterns.
+    static const std::string PrefixR;
+    /// Complete prefix for the left side of difference patterns.
+    static const std::string FullPrefixL;
+    /// Complete prefix for the right side of difference patterns.
+    static const std::string FullPrefixR;
+    /// Name for the function defining final instuction mapping.
+    static const std::string MappingFunctionName;
+    /// Name for pattern metadata nodes.
+    static const std::string MetadataName;
+    /// Set of loaded instruction difference patterns.
+    std::unordered_set<InstPattern> InstPatterns;
+    /// Set of loaded value difference patterns.
+    std::unordered_set<ValuePattern> ValuePatterns;
+
+    PatternSet(std::string ConfigPath);
+
+    ~PatternSet();
+
+    /// Retrives pattern metadata attached to the given instruction, returning
+    /// true for valid pattern metadata nodes.
+    bool getPatternMetadata(PatternMetadata &Metadata,
+                            const Instruction &Inst) const;
+
+  private:
+    /// Basic information about the final instruction mapping present on one
+    /// side of a pattern.
+    using MappingInfo = std::pair<const Instruction *, int>;
+
+    /// Settings applied to all pattern files.
+    StringMap<std::string> GlobalSettings;
+    /// Map of loaded pattern modules.
+    std::unordered_map<Module *, std::unique_ptr<Module>> PatternModules;
+    /// Map of loaded pattern module contexts.
+    std::unordered_map<Module *, std::unique_ptr<LLVMContext>> PatternContexts;
+
+    /// Load the given configuration file.
+    void loadConfig(std::string &ConfigPath);
+
+    /// Add a new difference pattern.
+    void addPattern(std::string &Path);
+
+    /// Finds the pattern type associated with the given pattern functions.
+    PatternType getPatternType(const Function *FnL, const Function *FnR);
+
+    /// Initializes a pattern, loading all metadata, start positions, and the
+    /// final instruction mapping.
+    bool initializeInstPattern(InstPattern &Pat);
```

Comment:
Because the initialization may fail. This seems better than constructor exceptions.

Additionally, I aimed for `Pattern`, `InstPattern` and `ValuePattern` to be simple structs, fully managed by `PatternSet` (though that is simply a design choice).